### PR TITLE
Addition of saveIntegrationPluginOfType request to SwaggerHubClient

### PR DESCRIPTION
**Pull request by michaelmelodysmartbear**
* Request will enable the creation of an SCM integration plugin within SwaggerHub via the plugin itself
* a SCM integration plugin will allow changes made in SwaggerHub to be pushed to the SCM of the users choice
* Test added to verify the new functionality added

### CODE REVIEW

File: src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    private String token;
+    private String repository;
+    private String repositoryOwner;
+    private String enabled;
+    private String branch;
+    private String syncMethod;
+    private String outputFile;
+    private String target;
+
+    public SaveSCMPluginConfigRequest(Builder builder) {
```
Comment (KevinNewman-SB) [inline src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java]: I think typically in Builder pattern you make the Constructor of the Object private, so you force callers to use the Builder... not saying this is wrong, but if it has a builder it should use it.

File: src/main/java/io/swagger/swaggerhub/plugin/requests/UploadRequest.java [hunk 2]
```diff
@@ -36,7 +36,7 @@ public boolean isPrivate() {
 
     public String getOas() { return oas; }
 
-    private SwaggerHubRequest(Builder builder) {
+    private UploadRequest(Builder builder) {
```
Comment (KevinNewman-SB) [inline src/main/java/io/swagger/swaggerhub/plugin/requests/UploadRequest.java]: ah! I see you have it private here, so you must have a reason for making it public above

File: src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        private String outputFile;
+        private String target;
+
+        public Builder(String apiOwner, String api, String version) {
+            this.api = api;
+            this.apiOwner = apiOwner;
+            this.version = version;
+        }
+
+        public SaveSCMPluginConfigRequest.Builder oas(String oas){
```
Comment (KevinNewman-SB) [inline src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java]: i think convention for builder pattern is like 

withOas
withScmProvidor
addToken ect....

Not super keen for or against it, just pointing it out.

File: src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    public String getRequestBody() throws JsonProcessingException {
+        return new ObjectMapper().writeValueAsString(new SCMIntegrationPluginConfiguration(this));
+    }
+
+    /**
+     * Data transfer object used to send a potential SCM plugin configuration as a request body on a POST/PUT request
+     */
+    public class SCMIntegrationPluginConfiguration {
```
Comment (KevinNewman-SB) [inline src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java]: why is this an inner class? I cant see why its not its own class

File: src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java (hunk 1 above)
Comment (michaelmelodysmartbear) [reply src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java]: Good spot! My error 👍

File: src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java (hunk 4 above)
Comment (michaelmelodysmartbear) [reply src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java]: Originally had this as part of a DTO's package but wanted to cut down on the size of the PR.

Comment (michaelmelodysmartbear) [reply src/main/java/io/swagger/swaggerhub/plugin/requests/SaveSCMPluginConfigRequest.java]: Moving out into dtos package.

**Comment by jsfrench** [comment]
This functionality is going to be exposed to users via the plugin config at some point I assume?

**Comment by michaelmelodysmartbear** [comment]
> Not sure that I understand why generic code that's shared between upload and download functionality has been refactored with an exclusive focus on uploading (e.g. `getBaseUrl` is now `getUploadUrl` and `SwaggerHubRequest`is now `UploadRequest`.
> 
> Ends up with code that downloads a definition from SwaggerHub having to use an `UploadRequest` like this, which seems a bit odd?
> 
> ```java
> String swaggerJson = swaggerHubClient.getDefinition(uploadRequest);
> ```
> 
> Downloading definitions from SwaggerHub was the main original use case and the main motiviation for building this plugin, customers wanted it and it solved a problem for us internally. Kind of looks like that functionality is being refactored out of existence.

Thanks for this John. Complete miss on my behalf. Far too eager when refactoring and lost sight of current existing functionality. Reverted back to the original naming.

**Comment by michaelmelodysmartbear** [comment]
> This functionality is going to be exposed to users via the plugin config at some point I assume?

Yep. That's the next PR to come for the end of the week. The idea being that once the user has uploaded a definition, they have the option of automagically creating an SCM integration plugin that will give them communication back to their SCM of choice.

**Comment by michaelmelodysmartbear** [comment]
Closing PR due to required changes being made in other branch.

**Pull request closed without merging**